### PR TITLE
lib: use mkImpureConfigMerger for runtime config merging

### PR DESCRIPTION
**Pull request by yajo**
### Description
Add a reusable `mkImpureConfigMerger` function to `lib.hm.generators`
that generates a bash script snippet for merging Nix-generated config
files with existing user config files at activation time. This preserves
any interactive changes the user has made while applying Nix-declared
settings on top.

The function uses `jaq` for both the merge operation and format
conversion, supporting JSON, YAML, TOML, and CBOR natively via its
`--from`/`--to` flags. An optional `reader` parameter allows custom
preprocessing (e.g., JSON5 with comments).

Refactor three existing modules to use the new library function:

- `programs.zed-editor`: Replaces local `impureConfigMerger` with
  `mkImpureConfigMerger`, preserving JSON5 reader support.
- `programs.obsidian`: Replaces inline jq merge script with
  `mkImpureConfigMerger`.
- `programs.joplin-desktop`: Replaces inline jq merge script with
  `mkImpureConfigMerger`.

Also add `jaq` to the test whitelist in `tests/default.nix` so the
activation scripts can resolve it at test time.

Assisted-by: OpenCode + DeepSeek v4 Flash

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `nix fmt` or
      `nix-shell -A dev --run treefmt`.

- [x] Code tested through `nix build .#test-all`
      or a targeted `nix run .#tests -- <pattern>`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

  ```
  {component}: {description}

  {long description}
  ```

  See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)
  - [ ] Basic tests added. See [Tests](https://nix-community.github.io/home-manager/index.xhtml#sec-tests)

- If this PR adds an exciting new feature or contains a breaking change.
  - [ ] Generate a news entry. See [News](https://nix-community.github.io/home-manager/index.xhtml#sec-news)

### CODE REVIEW

**Review by karaolidis — CHANGES_REQUESTED** [review]
Hey, I hate to be that guy, but I don't think this should be merged as-is:

- The new helper emits raw jaq/printf, so obsidian's merge now runs even under --dry-run and stops logging in verbose mode.
- Every zed/obsidian/joplin user now drags jaq into their activation closure, and jq is far more likely to already be there

zed and joplin should probably be updated to add dry-run and verbose mode, but I don't think a shared helper is the right approach for all three.

edit: this is my own opinion of course, I am open to counterarguments :)

**Comment by yajo** [comment]
I started by combining `jq` with `remarshal`, but honestly it felt like `jaq` is the tool for the job. Just one tool that does all we need.

Regarding dry-run, I didn't even know there was a dry-run mode! 😆 Let me check that.

**Comment by yajo** [comment]
I pushed support for dry run and verbose modes. Some tests included too.

Regarding the removal of jaq, I still feel like it's the right tool here...

**Comment by yajo** [comment]
FWIW I'm missing this in some downstream projects too, that might get upstreamed at some point.

For example, for configuring Goose, I want some settings to be static, such as MCP providers, coming from `programs.mcp`, while others dynamic such as the default provider and model (which change like every week). [The file is YAML in this case](https://goose-docs.ai/docs/guides/config-files/).

**Comment by karaolidis** [comment]
Hey, genuinely forgot to re-review this, my apologies! Give me 5 mins :)

File: modules/lib/generators.nix [hunk 1]
```diff
@@ -916,6 +916,128 @@ in
...
+          "printf '%s\\n' \"$config\" | ${jaqBin} --to ${format} -c '.' > ${lib.escapeShellArg path}";
+      defaultVerboseMsg = "Merging Nix-generated config into ${path}";
+      verboseMsg' = if verboseMsg != null then verboseMsg else defaultVerboseMsg;
+    in
+    ''
+      if [[ -v VERBOSE ]]; then
+        echo ${lib.escapeShellArg verboseMsg'}
+      fi
+      if [[ -v DRY_RUN ]]; then
+        echo "jaq -n ${lib.escapeShellArg jqOperation} --argjson dynamic ... --argjson static ..."
```
Comment (karaolidis) [inline modules/lib/generators.nix]: `escapeShellArg` will emit single quotes, which don't protect properly inside the outer double quotes, so $dynamic/$static expand, and `set -u` kills the activation because they're unset.

File: modules/lib/generators.nix [hunk 2]
```diff
@@ -916,6 +916,128 @@ in
...
+        return 0
```
Comment (karaolidis) [inline modules/lib/generators.nix]: I am fairly certain we are not supposed to `return` from activation scripts, instead we're supposed to use `run`? I think this will fail. See: https://home-manager.dev/manual/unstable/options/home-manager/home.html#opt-home.activation

File: modules/lib/generators.nix [hunk 3]
```diff
@@ -916,6 +916,128 @@ in
...
+      fi
+      mkdir -p $(dirname ${lib.escapeShellArg path})
```
Comment (karaolidis) [inline modules/lib/generators.nix]: I think the unquoted dirname will fail on Darwin (~/Library/Application Support)

File: modules/lib/generators.nix [hunk 4]
```diff
@@ -916,6 +916,128 @@ in
...
+      if [ ! -e ${lib.escapeShellArg path} ]; then
+        : > ${lib.escapeShellArg path}
```
Comment (karaolidis) [inline modules/lib/generators.nix]: This creates a missing file, then:

dynamic="$(jaq -c '.' path 2>/dev/null || echo '{}')"

jaq on an empty file exits 0 with no output, so the || fallback never fires and $dynamic is the empty string:

exit=0  output=[]  len=0
Error: failed to parse: byte offset 0: value expected (for value passed to `--argjson dynamic`)
exit=5

File: modules/lib/generators.nix [hunk 5]
```diff
@@ -916,6 +916,128 @@ in
...
+      fi
+      dynamic="$(${readerCmd} ${lib.escapeShellArg path} 2>/dev/null || echo ${lib.escapeShellArg empty})"
+      static="$(cat ${lib.escapeShellArg staticSettings})"
```
Comment (karaolidis) [inline modules/lib/generators.nix]: Should this also go through jaq instead of cat? Because this assumes it is well-formed json so the other formats won't work? Or am I misreading this?

File: modules/lib/generators.nix [hunk 6]
```diff
@@ -916,6 +916,128 @@ in
...
+          reader
+        else if isJson then
+          "${jaqBin} -c '.'"
+        else
+          "${jaqBin} --from ${format} -c '.'";
+      writerCmd =
+        if isJson then
+          "printf '%s\\n' \"$config\" > ${lib.escapeShellArg path}"
+        else
+          "printf '%s\\n' \"$config\" | ${jaqBin} --to ${format} -c '.' > ${lib.escapeShellArg path}";
```
Comment (karaolidis) [inline modules/lib/generators.nix]: This writes before jaq runs, so if jaq fails, we have clobbered the existing config.

File: modules/lib/generators.nix [hunk 7]
```diff
@@ -916,6 +916,128 @@ in
...
+
+      `path` (string)
+      : Absolute path to the config file on disk.
+
+      `staticSettings` (path or derivation)
+      : Path to the Nix-generated static config file, typically produced by
+        `format.generate`.
+
+      `reader` (string or null; optional)
+      : Shell command that reads the existing file from stdin and writes JSON
```
Comment (karaolidis) [inline modules/lib/generators.nix]: Not stdin, passes the path as an arg.

**Review by karaolidis — COMMENTED** [review]
Thanks for pushing the update! I did some testing and discovered a bunch of bugs. Left detailed comments on the respective lines.

On upstreaming, I'd still keep this in your own flake as a lib extension rather than putting it in Home Manager.

The bugs are all fixable; that's not the argument. The argument is that all three are the same failure: the helper emits bash that has to be correct against the activation contract (set -eu, top-level inlining, run/verboseEcho, dry-run, cd $HOME).

Upstreaming means following that contract for a {format} * {reader} * {missing, empty, valid, unparseable} * {live, dry-run} * {Linux, Darwin} matrix. And the merge semantics are also per-app (joplin wants shallow +, obsidian and zed want *, zed's keymaps need group_by/reduce).

The benefit is ~40 lines of de-duplication across three modules that already worked.

IMO, the genuinely valuable parts are separable and much lower risk: teach zed and joplin to use run/verboseEcho so they respect --dry-run and --verbose.

Still happy to be outvoted by another maintainer 🙃

**Comment by yajo** [comment]
Thanks for the review! I'll attend all those comments.

Regarding this:

> The benefit is ~40 lines of de-duplication across three modules that already worked.

I think it's a bit short-scoped. This problem of "apps that only support a single configuration file but need write access to it to be useful" is more common than just Joplin, Zed and Obsidian.

Having this library function will help:
- Make it easier to add new apps.
- Make it easier to maintain current apps (having code already triplicated is a burden).

I have no vote here, but honestly I think this is a very helpful enhancement.

**Comment by yajo** [comment]
Thanks for the second round — the line comments were spot on, and the bugs were real. All of them are fixed now:

- Removed the top-level `return` in the dry-run branch; the snippet now wraps the whole merge in `if [[ -v DRY_RUN ]]; then … else … fi`.
- Quoted `$(dirname …)` so paths with spaces work on Darwin.
- Guarded `$dynamic` for the empty-but-readable file case (jaq exits 0 with no output, so the `|| echo` fallback never fired).
- Missing files are initialized with the `empty` value instead of a blank file.
- The merge result is written to a temp file and `install`ed only on success, so a failing merge can't clobber the existing config.
- Fixed the `reader` docs (the file is passed as the last argument, not read from stdin).

I also added a test that *executes* the generated snippet under `DRY_RUN` and `VERBOSE` (`tests/modules/programs/zed-editor/settings-dry-run.nix`): dry-run leaves the file untouched and doesn't trip over unset variables, verbose logs the merge message, quiet mode logs nothing. The full generators/obsidian/joplin/zed suite passes.

On keeping the helper in the library, I'd like to push back a bit.

I actually read your bug list as the strongest argument *for* a shared helper, not against it. The activation contract — `set -eu`, top-level inlining, `DRY_RUN`/`VERBOSE`, `cd $HOME` — is subtle, and I got a good chunk of it wrong in the first version. The point of centralizing it is that this contract gets implemented and reviewed **once**, and every module that merges configs inherits the fixes. If each module keeps its own script (or we just bolt `run`/`verboseEcho` onto zed and joplin), the next module that needs config merging will re-derive the same edge cases — empty files, spaces in paths, clobbering on failure — and each one will get a different subset of them right. That's the same class of bug, N times over.

On the matrix: I think it's smaller than it looks, because the per-app merge semantics aren't part of it — they're the `jqOperation` parameter (joplin's shallow `+`, obsidian/zed's `*`, zed's `group_by` keymaps are all passed in, not hardcoded). What's left is {json, other formats} × {missing/empty, valid, unparseable} × {live, dry-run}, and "missing" collapses into the same `empty` fallback as "empty". That's a handful of cases, and the new test covers the ones that actually bit us.

Regarding jaq, as said I still believe it's the right choice, but we can improve the situation. Some options:
- Remove jaq and use jq. Then, pipe with remarshal when other formats support is needed. This would complicate a bit the function (but not extremely). The extra bits would be only for those that really need remarshal.
- Remove jq and use only jaq elsewhere. jaq is still a bit bigger than jq, but it is also faster and works with more formats, so it's also a good option to avoid duplicates.
- Keep it as it is right now.

In short: I still think a reviewed-once `mkImpureConfigMerger` is more robust than three hand-rolled scripts each re-negotiating the activation contract.

**Review by zh4ngx — CHANGES_REQUESTED** [review]
Thanks for working on this. Centralizing the merge logic for Zed, Obsidian, and Joplin makes sense since all three already need write-capable configs and hand-rolled snippets are easy to get wrong.

Before this lands, there are two data-integrity issues in the activation snippet that need fixing:

1. **Silent config loss on read error (line 1037)**:
   `dynamic=$(readerCmd path 2>/dev/null || echo empty)`
   If an existing config file is corrupted, unreadable, or invalid JSON, the `|| echo empty` fallback treats it as an absent file and line 1043 installs the default generated config over it. This silently wipes existing user settings. An existing file that fails to parse must fail activation loudly rather than falling back to `empty`.

2. **Unconditional mode overwrite (line 1043)**:
   `install -m644` rewrites the file permissions regardless of what the user or application originally set. The snippet should preserve the existing file mode.

On architecture, I agree with keeping this as an **internal shared helper** across the modules that need it rather than exposing it as a public `lib.hm.generators` API. A public generator promises formats (YAML, TOML, CBOR) that currently lack execution tests against the activation contract. Keeping it internal first avoids locking in the public API before the edge cases are proven in practice.

**Comment by yajo** [comment]
Thanks, issues attended.



> keeping this as an **internal shared helper**

I actually need this in other projects, so I prefer to have it reusable from the outside. However, I marked it as experimental in the docs, so there's no stability guarantee for now.
